In [1]:
import pandas as pd
import io
import requests

In [2]:
aprilurl="https://media.githubusercontent.com/media/DataScienceNishant/Datafiles/master/Voice%20Quality/CallVoiceQualityExperience-2018-April.csv"

mayurl="https://media.githubusercontent.com/media/DataScienceNishant/Datafiles/master/Voice%20Quality/CallVoiceQuality_Data_2018_May.csv"

In [4]:
aprilrqst=requests.get(aprilurl).content
april=pd.read_csv(io.StringIO(aprilrqst.decode("utf-8")))

mayrqst=requests.get(mayurl).content
may=pd.read_csv(io.StringIO(mayrqst.decode("utf-8")))

april

Operator Indoor_Outdoor_Travelling Network Type  Rating  \
0        Airtel                    Indoor           3G       5   
1          RJio                    Indoor           4G       4   
2        Airtel                   Outdoor           3G       5   
3        Airtel                Travelling           3G       5   
4          RJio                    Indoor           4G       5   
5          RJio                Travelling           4G       4   
6        Airtel                Travelling           4G       1   
7          RJio                   Outdoor           4G       4   
8          RJio                    Indoor           4G       5   
9          RJio                    Indoor           4G       2   
10         Idea                   Outdoor      Unknown       4   
11         RJio                    Indoor           4G       5   
12       Airtel                   Outdoor           4G       5   
13         BSNL                    Indoor           3G       1   
14         RJio                Travelling           4G       5   
15     Vodafone                   Outdoor      Unknown       5   
16         RJio                Travelling           4G       3   
17       Airtel                    Indoor           4G       1   
18       Airtel                    Indoor           4G       1   
19         RJio                    Indoor           4G       1   
20         RJio                    Indoor           4G       1   
21         RJio                    Indoor           4G       1   
22       Airtel                    Indoor           3G       1   
23         RJio                    Indoor           4G       1   
24         RJio                    Indoor      Unknown       1   
25         BSNL                   Outdoor      Unknown       4   
26         RJio                    Indoor           4G       2   
27     Vodafone                    Indoor           3G       4   
28       Airtel                    Indoor           3G       1   
29         RJio                    Indoor           4G       4   
...         ...                       ...          ...     ...   
63306      BSNL                    Indoor           4G       5   
63307      BSNL                    Indoor           4G       5   
63308      BSNL                    Indoor           4G       5   
63309      BSNL                   Outdoor           4G       4   
63310      BSNL                   Outdoor           4G       4   
63311      BSNL                Travelling           4G       5   
63312      Idea                Travelling           4G       4   
63313    Airtel                    Indoor           3G       5   
63314      RJio                    Indoor           4G       1   
63315      RJio                    Indoor           4G       1   
63316      RJio                    Indoor           4G       2   
63317    Airtel                    Indoor      Unknown       5   
63318      RJio                    Indoor           4G       4   
63319    Airtel                   Outdoor           3G       4   
63320    Airtel                   Outdoor           3G       1   
63321      RJio                    Indoor           4G       4   
63322      RJio                    Indoor           4G       4   
63323      RJio                    Indoor           4G       5   
63324      RJio                    Indoor           4G       1   
63325      RJio                    Indoor           4G       1   
63326      RJio                    Indoor           4G       1   
63327      RJio                    Indoor           4G       1   
63328    Airtel                   Outdoor      Unknown       5   
63329      RJio                    Indoor           4G       5   
63330  Vodafone                    Indoor           4G       5   
63331      RJio                    Indoor           4G       5   
63332    Airtel                    Indoor      Unknown       3   
63333    Airtel                Travelling      Unknown       4   
63334  Vodafone                    Indoor      

In [4]:
print(may["State Name"].unique())
print(april["State Name"].unique())

[nan 'NCT' 'Bihar' 'Chhattisgarh' 'Karnataka' 'Haryana' 'Andhra Pradesh'
 'Telangana' 'Uttar Pradesh' 'Tamil Nadu' 'Odisha' 'Maharashtra'
 'Uttarakhand' 'Gujarat' 'Himachal Pradesh' 'Jharkhand' 'West Bengal'
 'Chandigarh' 'Madhya Pradesh' 'Rajasthan' 'Kerala' 'Punjab' 'Goa' 'Assam'
 'Kashmir' 'Dadra and Nagar Haveli' 'Samdrup Jongkhar' 'New York'
 'Chhukha' 'Pondicherry' 'Samchi' 'Andaman and Nicobar Islands'
 'Arunachal Pradesh' 'Tripura' 'Eastern Region']
['Haryana' 'Tamil Nadu' 'Bihar' 'Andhra Pradesh' 'Maharashtra' 'Rajasthan'
 'Odisha' nan 'Uttar Pradesh' 'Karnataka' 'NCT' 'West Bengal'
 'Madhya Pradesh' 'Uttarakhand' 'Punjab' 'Gujarat' 'Nagaland' 'Jharkhand'
 'Telangana' 'Kerala' 'Assam' 'Chhattisgarh' 'Himachal Pradesh' 'Chhukha'
 'Pondicherry' 'Goa' 'Chandigarh' 'Kashmir' 'Eastern Region' 'Sikkim'
 'Meghalaya' 'Manipur' 'Arunachal Pradesh' 'Adis Abeba'
 'Dadra and Nagar Haveli' 'Samchi' 'Tripura' 'Central Region']


In [5]:
april[april["Network Type"]=='Unknown']

Operator Indoor_Outdoor_Travelling Network Type  Rating  \
10         Idea                   Outdoor      Unknown       4   
15     Vodafone                   Outdoor      Unknown       5   
24         RJio                    Indoor      Unknown       1   
25         BSNL                   Outdoor      Unknown       4   
32         Idea                    Indoor      Unknown       4   
...         ...                       ...          ...     ...   
63286  Vodafone                    Indoor      Unknown       5   
63317    Airtel                    Indoor      Unknown       5   
63328    Airtel                   Outdoor      Unknown       5   
63332    Airtel                    Indoor      Unknown       3   
63333    Airtel                Travelling      Unknown       4   

      Call Drop Category   Latitude  Longitude   State Name  
10          Satisfactory  -1.000000  -1.000000          NaN  
15          Satisfactory  -1.000000  -1.000000          NaN  
24          Poor Network  -1.000000  -1.000000          NaN  
25          Satisfactory  -1.000000  -1.000000          NaN  
32          Satisfactory  19.229211  72.859561  Maharashtra  
...                  ...        ...        ...          ...  
63286       Satisfactory  22.629600  88.380250  West Bengal  
63317       Satisfactory  25.233757  86.963704        Bihar  
63328       Satisfactory  24.433787  85.530784    Jharkhand  
63332       Poor Network  19.408291  72.838735  Maharashtra  
63333       Satisfactory  19.408291  72.838735  Maharashtra  

[12099 rows x 8 columns]

In [6]:
# aprilnull=april.dropna()

In [7]:
april["Operator"].unique()

array(['Airtel', 'RJio', 'Idea', 'BSNL', 'Vodafone', 'Tata', 'MTNL',
       'RComm', 'Telenor', 'Other'], dtype=object)

In [8]:
# april.isnull().sum()
# april[(april["Operator"]=='Airtel') & (april["Network Type"]=='3G') ][["Network Type","Rating","Call Drop Category","State Name"]]
# april.groupby("State Name")
# aprilsorted=april.sort_index(by=["State Name","Network Type"])
# aprilsorted
newapril=april
newmay=may
# newapril["Network Type"]=newapril["Network Type"].apply(lambda x: '4G' if x == 'Unknown' else x)
# newapril

In [9]:
# april[april["Indoor_Outdoor_Travelling"]=='Outdoor'].sort_index(by="State Name")[["Network Type","Indoor_Outdoor_Travelling","State Name"]].head(100)
aprilgb=april.groupby(["Operator","Network Type"]).count()
# maygb=may.groupby(["Operator","Network Type"]).count()
# aprilgb[aprilgb["State Name"]=='Arunachal Pradesh']
aprilgb.head(60)
# aprilgb

Indoor_Outdoor_Travelling  Rating  Call Drop Category  \
Operator Network Type                                                          
Airtel   2G                                  951     951                 951   
         3G                                 5183    5183                5183   
         4G                                 4876    4876                4876   
         Unknown                            3889    3889                3889   
BSNL     2G                                 1454    1454                1454   
         3G                                 2973    2973                2973   
         4G                                  455     455                 455   
         Unknown                            3202    3202                3202   
Idea     2G                                  430     430                 430   
         3G                                 1944    1944                1944   
         4G                                 1438    1438                1438   
         Unknown                            1922    1922                1922   
MTNL     2G                                   18      18                  18   
         3G                                   41      41                  41   
         4G                                    2       2                   2   
         Unknown                              81      81                  81   
Other    3G                                   24      24                  24   
         Unknown                              12      12                  12   
RComm    2G                                  210     210                 210   
         3G                                    7       7                   7   
RJio     4G                                24938   24938               24938   
         Unknown                             618     618                 618   
Tata     2G                                    1       1                   1   
         3G                                   91      91                  91   
         4G                                   12      12                  12   
         Unknown                              93      93                  93   
Telenor  2G                                    2       2                   2   
Vodafone 2G                                  742     742                 742   
         3G                                 3330    3330                3330   
         4G                                 2115    2115                2115   
         Unknown                            2282    2282                2282   

                       Latitude  Longitude  State Name  
Operator Network Type                                   
Airtel   2G                 951        951         474  
         3G                5183       5183        3630  
         4G                4876       4876        3873  
         Unknown           3889       3889        3050  
BSNL     2G                1454       1454        1219  
         3G                2973       2973        2033  
         4G                 455        455         365  
         Unknown           3202       3202        2106  
Idea     2G                 430        430         284  
         3G                1944       1944        1458  
         4G                1438       1438        1091  
         Unknown           1922       1922        1341  
MTNL     2G                  18         18          13  
         3G                  41         41          33  
         4G                   2          2           1  
         Unknown             81         81          64  
Other    3G                  24         24          24  
         Unknown             12         12           1  
RComm    2G                 210        210         209  
         3G                   7          7           7  
RJio     4G               24938      24938       18907  
         Unknown            618        618         441  
Tata     2G   

<b><h1>From this data it is clear that <i>"unknown"<i> network type in :</h1><br>
    <h3>"Airtel" comes from 4G<br>
    "BSNL"   comes from 3G<br>
    "Idea"   comes from 3G<br>
    "MTNL"   comes from 2G<br>
    "Other"  comes from 2G<br>
    "RJio"   comes from 4G<br>
    "Tata"   comes from 3G<br>
    "Vodafone" comes from 3G<br>

In [10]:
# def runknown(oprtr,ntwrk):
#     if ntwrk=='Unknown':
#         if oprtr=='Airtel' or oprtr=='RJio':
#             ntwrk='4G'
#         elif oprtr=='BSNL' or oprtr=='Idea' or oprtr=='Tata' or oprtr=='Vodafone':
#             ntwrk='3G'
#         else:
#             ntwrk='2G'
    
newapril=april
newapril.loc[((newapril["Operator"] == 'Airtel') |  (newapril["Operator"] == 'RJio')) & (newapril["Network Type"]=='Unknown'), "Network Type"] = '4G'  
newapril.loc[((newapril["Operator"] == 'MTNL') |  (newapril["Operator"] == 'Other')) & (newapril["Network Type"]=='Unknown'), "Network Type"] = '2G'  
newapril.loc[(newapril["Network Type"]=='Unknown'), "Network Type"] = '2G'  

newapril.fillna("",inplace=True)
newapril

Operator Indoor_Outdoor_Travelling Network Type  Rating  \
0        Airtel                    Indoor           3G       5   
1          RJio                    Indoor           4G       4   
2        Airtel                   Outdoor           3G       5   
3        Airtel                Travelling           3G       5   
4          RJio                    Indoor           4G       5   
...         ...                       ...          ...     ...   
63331      RJio                    Indoor           4G       5   
63332    Airtel                    Indoor           4G       3   
63333    Airtel                Travelling           4G       4   
63334  Vodafone                    Indoor           4G       4   
63335      RJio                    Indoor           4G       5   

      Call Drop Category   Latitude  Longitude   State Name  
0           Satisfactory  28.422966  76.912324      Haryana  
1           Satisfactory  11.158358  77.301897   Tamil Nadu  
2           Satisfactory  28.422931  76.912253      Haryana  
3           Satisfactory  28.422947  76.912260      Haryana  
4           Satisfactory  25.625990  85.094294        Bihar  
...                  ...        ...        ...          ...  
63331       Satisfactory  26.855378  89.375578  West Bengal  
63332       Poor Network  19.408291  72.838735  Maharashtra  
63333       Satisfactory  19.408291  72.838735  Maharashtra  
63334       Satisfactory  30.323117  78.001541  Uttarakhand  
63335       Satisfactory  -1.000000  -1.000000               

[63336 rows x 8 columns]

In [11]:
    
newmay=may
newmay.loc[((newmay["Operator"] == 'Airtel') |  (newmay["Operator"] == 'RJio')) & (newmay["Network Type"]=='Unknown'), "Network Type"] = '4G'  
newmay.loc[((newmay["Operator"] == 'MTNL') |  (newmay["Operator"] == 'Other')) & (newmay["Network Type"]=='Unknown'), "Network Type"] = '2G'  
newmay.loc[(newmay["Network Type"]=='Unknown'), "Network Type"] = '2G'  

newmay.fillna("",inplace=True)
newmay

Operator In Out Travelling Network Type  Rating  Call Drop Category  \
0         RJio        Travelling           4G       5        Satisfactory   
1         RJio        Travelling           4G       2  Poor Voice Quality   
2         BSNL            Indoor           3G       5        Satisfactory   
3         BSNL            Indoor           3G       5        Satisfactory   
4         BSNL            Indoor           3G       5        Satisfactory   
...        ...               ...          ...     ...                 ...   
31976     RJio            Indoor           4G       4        Satisfactory   
31977   Airtel            Indoor           4G       5        Satisfactory   
31978     RJio            Indoor           4G       5        Satisfactory   
31979     RJio            Indoor           4G       5        Satisfactory   
31980     BSNL            Indoor           2G       5        Satisfactory   

        Latitude  Longitude   State Name  
0      -1.000000  -1.000000               
1      28.593588  77.164276          NCT  
2      -1.000000  -1.000000               
3      -1.000000  -1.000000               
4      -1.000000  -1.000000               
...          ...        ...          ...  
31976  20.979739  75.580521  Maharashtra  
31977  17.438340  78.382000    Telangana  
31978  -1.000000  -1.000000               
31979  28.533182  77.216453          NCT  
31980  17.406222  78.438804    Telangana  

[31981 rows x 8 columns]

In [12]:
print(newapril.columns)
print(newmay.columns)

Index(['Operator', 'Indoor_Outdoor_Travelling', 'Network Type', 'Rating',
       'Call Drop Category', 'Latitude', 'Longitude', 'State Name'],
      dtype='object')
Index(['Operator', 'In Out Travelling', 'Network Type', 'Rating',
       'Call Drop Category', 'Latitude', 'Longitude', 'State Name'],
      dtype='object')


In [13]:
aprilme=newapril
# aprilme["Network"]=newapril["Operator"]+" "+newapril["Network Type"]
aprilme.drop(["Latitude","Longitude"],axis=1,inplace=True)

mayme=newmay
mayme.drop(["Latitude","Longitude"],axis=1,inplace=True)
# mayme["Network"]=newmay["Operator"]+" "+newmay["Network Type"]

In [14]:
# def columns(Dataframe):
#     df=Dataframe
#     cols=list(df.columns)
#     cols=[cols[-1]]+cols[:-1]
#     return df[cols]

# aprilme=columns(aprilme)
aprilme.columns=['Operator', 'In Out Travelling', 'Network Type', 'Rating','Call Drop Category','State Name']
mayme.columns=['Operator', 'In Out Travelling', 'Network Type', 'Rating','Call Drop Category','State Name']
aprilme["Month"]="April"
mayme["Month"]="May"

In [15]:
callqlty=pd.concat([aprilme,mayme])
callqlty

Operator In Out Travelling Network Type  Rating Call Drop Category  \
0       Airtel            Indoor           3G       5       Satisfactory   
1         RJio            Indoor           4G       4       Satisfactory   
2       Airtel           Outdoor           3G       5       Satisfactory   
3       Airtel        Travelling           3G       5       Satisfactory   
4         RJio            Indoor           4G       5       Satisfactory   
...        ...               ...          ...     ...                ...   
31976     RJio            Indoor           4G       4       Satisfactory   
31977   Airtel            Indoor           4G       5       Satisfactory   
31978     RJio            Indoor           4G       5       Satisfactory   
31979     RJio            Indoor           4G       5       Satisfactory   
31980     BSNL            Indoor           2G       5       Satisfactory   

        State Name  Month  
0          Haryana  April  
1       Tamil Nadu  April  
2          Haryana  April  
3          Haryana  April  
4            Bihar  April  
...            ...    ...  
31976  Maharashtra    May  
31977    Telangana    May  
31978                 May  
31979          NCT    May  
31980    Telangana    May  

[95317 rows x 7 columns]

In [22]:
callqlty.groupby("Network Type").count()

Operator  In Out Travelling  Rating  Call Drop Category  \
Network Type                                                            
2G               16723              16723   16723               16723   
3G               19151              19151   19151               19151   
4G               59443              59443   59443               59443   

              State Name  Month  
Network Type                     
2G                 16723  16723  
3G                 19151  19151  
4G                 59443  59443

<b> So according to this report<br>
<h1>
2G users all over the india are around "16723"<br>
3G users all over the india are around "19151"<br>
    4G users all over the india are around "59443"<br></h1>

In [69]:
usersnum=callqlty.groupby(["Operator","Network Type"],as_index=False).count()
# usersnum[(usersnum["Network Type"]=='2G') & (usersnum["Month"]==usersnum[usersnum["Network Type"]=='2G'].max())]
# usersnum
# print(usersnum[usersnum["Network Type"]=='2G'].max()[0])
# print(usersnum[usersnum["Network Type"]=='3G'].max()[0])
# print(usersnum[usersnum["Network Type"]=='4G'].max()[0])
fG=usersnum[usersnum["Network Type"]=='4G']
twG=usersnum[usersnum["Network Type"]=='2G']
thG=usersnum[usersnum["Network Type"]=='3G']
print(fG[fG["Rating"]==fG["Rating"].max()]["Operator"]+" has more no of 4G users")
print(thG[thG["Rating"]==thG["Rating"].max()]["Operator"]+" has more no of 3G users")
print(twG[twG["Rating"]==twG["Rating"].max()]["Operator"]+" has more no of 2G users")

16    RJio has more no of jio operators
Name: Operator, dtype: object
1    Airtel has more no of jio operators
Name: Operator, dtype: object
3    BSNL has more no of jio operators
Name: Operator, dtype: object


<b>No of users in different comapny<br<h1>
<i>Jio</i> has more no of 4G users<br>
<i>Airtel</i> has more no of 3G users<br>
<i>BSNL</i> has more no of 2G users<br>

In [98]:
callqlty.groupby("State Name").sum()["Rating"].nlargest(2)

State Name
               83306
Maharashtra    47349
Name: Rating, dtype: int64

<b> from above output it is cleared that<br>
    <h1><i>Maharashtra</i> has the best call quality ratings</h1>

In [89]:
callqlty.groupby(["Operator","In Out Travelling","Call Drop Category"]).count()

Network Type  Rating  \
Operator In Out Travelling Call Drop Category                         
Airtel   Indoor            Call Dropped                1435    1435   
                           Poor Network                2264    2264   
                           Poor Voice Quality          1353    1353   
                           Satisfactory                9779    9779   
         Outdoor           Call Dropped                 438     438   
...                                                     ...     ...   
Vodafone Outdoor           Satisfactory                2325    2325   
         Travelling        Call Dropped                 128     128   
                           Poor Network                 118     118   
                           Poor Voice Quality           127     127   
                           Satisfactory                1088    1088   

                                               State Name  Month  
Operator In Out Travelling Call Drop Category                     
Airtel   Indoor            Call Dropped              1435   1435  
                           Poor Network              2264   2264  
                           Poor Voice Quality        1353   1353  
                           Satisfactory              9779   9779  
         Outdoor           Call Dropped               438    438  
...                                                   ...    ...  
Vodafone Outdoor           Satisfactory              2325   2325  
         Travelling        Call Dropped               128    128  
                           Poor Network               118    118  
                           Poor Voice Quality         127    127  
                           Satisfactory              1088   1088  

[99 rows x 4 columns]

In [91]:
callqlty[callqlty["Operator"]=='RJio']

Operator In Out Travelling Network Type  Rating Call Drop Category  \
1         RJio            Indoor           4G       4       Satisfactory   
4         RJio            Indoor           4G       5       Satisfactory   
5         RJio        Travelling           4G       4       Satisfactory   
7         RJio           Outdoor           4G       4       Satisfactory   
8         RJio            Indoor           4G       5       Satisfactory   
...        ...               ...          ...     ...                ...   
31972     RJio            Indoor           4G       4       Satisfactory   
31975     RJio            Indoor           4G       4       Satisfactory   
31976     RJio            Indoor           4G       4       Satisfactory   
31978     RJio            Indoor           4G       5       Satisfactory   
31979     RJio            Indoor           4G       5       Satisfactory   

           State Name  Month  
1          Tamil Nadu  April  
4               Bihar  April  
5      Andhra Pradesh  April  
7           Rajasthan  April  
8              Odisha  April  
...               ...    ...  
31972     West Bengal    May  
31975                    May  
31976     Maharashtra    May  
31978                    May  
31979             NCT    May  

[39220 rows x 7 columns]

In [103]:
callqlty.groupby("State Name").mean().iloc[1:,:]

Rating
State Name                           
Adis Abeba                   3.300000
Andaman and Nicobar Islands  1.000000
Andhra Pradesh               3.448439
Arunachal Pradesh            2.272727
Assam                        4.448665
Bihar                        3.168323
Central Region               2.875000
Chandigarh                   4.696629
Chhattisgarh                 3.706480
Chhukha                      2.714286
Dadra and Nagar Haveli       3.450000
Eastern Region               2.750000
Goa                          2.995370
Gujarat                      3.403222
Haryana                      3.332602
Himachal Pradesh             4.205231
Jharkhand                    3.600000
Karnataka                    3.623348
Kashmir                      4.347107
Kerala                       2.817797
Madhya Pradesh               3.353619
Maharashtra                  3.396872
Manipur                      1.000000
Meghalaya                    1.200000
NCT                          3.139039
Nagaland                     3.583333
New York                     3.000000
Odisha                       3.869776
Pondicherry                  3.000000
Punjab                       4.053968
Rajasthan                    2.944495
Samchi                       5.000000
Samdrup Jongkhar             3.000000
Sikkim                       3.142857
Tamil Nadu                   3.681703
Telangana                    3.590237
Tripura                      5.000000
Uttar Pradesh                3.694566
Uttarakhand                  3.665083
West Bengal                  3.613997

In [118]:
poornetwork=callqlty.groupby("Operator",as_index=False).sum()
poornetwork[poornetwork["Rating"]==poornetwork["Rating"].min()]

Operator  Rating
8  Telenor     125

In [120]:
poornetwork[poornetwork["Rating"]==poornetwork["Rating"].max()]

Operator  Rating
6     RJio  139737

<h3>No we can't say, as above output shows that Jio has highest number of positive Rating as compared to any another company

<h3>No we can't say that Jio provides worst service everywhere. Its service are betterat some parts of the country and worst at my home.

In [122]:
callqlty.groupby(["Operator","Call Drop Category"]).count()

In Out Travelling  Network Type  Rating  \
Operator Call Drop Category                                            
Airtel   Call Dropped                     2129          2129    2129   
         Poor Network                     3236          3236    3236   
         Poor Voice Quality               2080          2080    2080   
         Satisfactory                    15095         15095   15095   
BSNL     Call Dropped                      644           644     644   
         Poor Network                     1177          1177    1177   
         Poor Voice Quality                986           986     986   
         Satisfactory                     8805          8805    8805   
Idea     Call Dropped                      646           646     646   
         Poor Network                      987           987     987   
         Poor Voice Quality                585           585     585   
         Satisfactory                     6341          6341    6341   
MTNL     Call Dropped                        3             3       3   
         Poor Network                       22            22      22   
         Satisfactory                      119           119     119   
Other    Call Dropped                        2             2       2   
         Poor Network                        5             5       5   
         Poor Voice Quality                  4             4       4   
         Satisfactory                       34            34      34   
RComm    Call Dropped                       17            17      17   
         Poor Network                      191           191     191   
         Poor Voice Quality                  2             2       2   
         Satisfactory                        9             9       9   
RJio     Call Dropped                     2684          2684    2684   
         Poor Network                     3983          3983    3983   
         Poor Voice Quality               3043          3043    3043   
         Satisfactory                    29510         29510   29510   
Tata     Call Dropped                       32            32      32   
         Poor Network                       38            38      38   
         Poor Voice Quality                 22            22      22   
         Satisfactory                      329           329     329   
Telenor  Poor Voice Quality                 34            34      34   
         Satisfactory                       13            13      13   
Vodafone Call Dropped                     1266          1266    1266   
         Poor Network                     1618          1618    1618   
         Poor Voice Quality                904           904     904   
         Satisfactory                     8722          8722    8722   

                             State Name  Month  
Operator Call Drop Category                     
Airtel   Call Dropped              2129   2129  
         Poor Network              3236   3236  
         Poor Voice Quality        2080   2080  
         Satisfactory             15095  15095  
BSNL     Call Dropped               644    644  
         Poor Network              1177   1177  
         Poor Voice Quality         986    986  
         Satisfactory              8805   8805  
Idea     Call Dropped               646    646  
         Poor Network               987    987  
         Poor Voice Quality         585    585  
         Satisfactory              6341   6341  
MTNL     Call Dropped                 3      3  
         Poor Network                22     22  
         Satisfactory               119    119  
Other    Call Dropped                 2      2  
         Poor Network                 5      5  
         Poor Voice Quality           4      4  
         Satisfactory                34     34  
RComm    Call Dropped                17     17  
         Poor Network               191    191  
         Poor Voice Quality           2      2  
         Satisfactory                 9     

<h1>No, as above mentioned output clearly shows that Jio has the highest number of satisfactory calls</h1>

In [124]:
callqlty.groupby(["Network Type","Call Drop Category"]).count()

Operator  In Out Travelling  Rating  \
Network Type Call Drop Category                                        
2G           Call Dropped            1178               1178    1178   
             Poor Network            2205               2205    2205   
             Poor Voice Quality      1423               1423    1423   
             Satisfactory           11917              11917   11917   
3G           Call Dropped            1827               1827    1827   
             Poor Network            2705               2705    2705   
             Poor Voice Quality      1566               1566    1566   
             Satisfactory           13053              13053   13053   
4G           Call Dropped            4418               4418    4418   
             Poor Network            6347               6347    6347   
             Poor Voice Quality      4671               4671    4671   
             Satisfactory           44007              44007   44007   

                                 State Name  Month  
Network Type Call Drop Category                     
2G           Call Dropped              1178   1178  
             Poor Network              2205   2205  
             Poor Voice Quality        1423   1423  
             Satisfactory             11917  11917  
3G           Call Dropped              1827   1827  
             Poor Network              2705   2705  
             Poor Voice Quality        1566   1566  
             Satisfactory             13053  13053  
4G           Call Dropped              4418   4418  
             Poor Network              6347   6347  
             Poor Voice Quality        4671   4671  
             Satisfactory             44007  44007

<h1>NO, from above table it is clear that voice calls are better in 4G data as compared to 3G data</h1>

In [133]:
calldrop=callqlty[callqlty["Call Drop Category"]=='Call Dropped']
calldrop

Operator In Out Travelling Network Type  Rating Call Drop Category  \
81       Airtel            Indoor           2G       1       Call Dropped   
94       Airtel            Indoor           3G       1       Call Dropped   
97         RJio            Indoor           4G       2       Call Dropped   
107    Vodafone           Outdoor           4G       1       Call Dropped   
114        Idea            Indoor           4G       2       Call Dropped   
...         ...               ...          ...     ...                ...   
31908    Airtel            Indoor           4G       1       Call Dropped   
31913  Vodafone           Outdoor           4G       1       Call Dropped   
31957      RJio           Outdoor           4G       1       Call Dropped   
31958    Airtel            Indoor           3G       1       Call Dropped   
31966      BSNL            Indoor           3G       1       Call Dropped   

           State Name  Month  
81                     April  
94                     April  
97         Tamil Nadu  April  
107             Bihar  April  
114    Madhya Pradesh  April  
...               ...    ...  
31908          Punjab    May  
31913                    May  
31957                    May  
31958                    May  
31966                    May  

[7423 rows x 7 columns]

In [135]:
calldrop.groupby(["State Name","Operator"]).count().tail(60)

In Out Travelling  Network Type  Rating  \
State Name       Operator                                            
Madhya Pradesh   RJio                     48            48      48   
                 Tata                      1             1       1   
Maharashtra      Airtel                  142           142     142   
                 BSNL                     21            21      21   
                 Idea                    223           223     223   
                 MTNL                      2             2       2   
                 RComm                    14            14      14   
                 RJio                    907           907     907   
                 Tata                     16            16      16   
                 Vodafone                188           188     188   
Manipur          BSNL                      1             1       1   
Meghalaya        RJio                      2             2       2   
NCT              Airtel                   76            76      76   
                 BSNL                      3             3       3   
                 Idea                      7             7       7   
                 RJio                    128           128     128   
                 Vodafone                 55            55      55   
Odisha           Airtel                   23            23      23   
                 BSNL                     30            30      30   
                 RJio                     15            15      15   
Pondicherry      Vodafone                  1             1       1   
Punjab           Airtel                   19            19      19   
                 BSNL                      9             9       9   
                 Idea                     11            11      11   
                 RJio                      1             1       1   
                 Vodafone                  1             1       1   
Rajasthan        Airtel                   22            22      22   
                 BSNL                     15            15      15   
                 Idea                      1             1       1   
                 RJio                     74            74      74   
                 Vodafone                 41            41      41   
Samdrup Jongkhar BSNL                      1             1       1   
Sikkim           Vodafone                  2             2       2   
Tamil Nadu       Airtel                   99            99      99   
                 BSNL                     50            50      50   
                 Idea                      2             2       2   
                 RJio                     67            67      67   
                 Vodafone                155           155     155   
Telangana        Airtel                  187           187     187   
                 BSNL                      8             8       8   
                 Idea                     20            20      20   
                 RJio                     85            85      85   
                 Tata                      1             1       1   
                 Vodafone                 11            11      11   
Uttar Pradesh    Airtel                  158           158     158   
                 BSNL                     24            24      24   
                 Idea                     26            26      26   
                 RJio                     96            96      96   
                 Tata                      1             1       1   
                 Vodafone                 79            79      79   
Uttarakhand      Airtel                   17            17      17   
                 BSNL                      6             6       6   
                 Idea                      2             2       2   
                 RJio                      4             4       4   
                 Vodafone                  2             2       2   
West Bengal      Airtel              

<h1>No, we can't say as in each state different operators have highest call drops</h1>